### PR TITLE
KAFKA-17502: Modified commitSync() and close() handling in clients

### PR DESCRIPTION
**Pull request by ShivsundarR**
*What*
Currently the code in `ShareConsumeRequestManager` works on the basis that there can only be one `commitSync()`/`close()` at a time. But there is a chance these calls timeout on the application thread, but are still sent later on the background thread. This will mean the incoming commitSync()/close() will not be processed, resulting in possible loss of acknowledgements.

To cover this case, we will now have a list of `AcknowledgeRequestStates` to store the `commitSyncs()` and a separate requestState to store the `close()`. This queue will be processed one by one until its empty. For `close()`, we are still assuming there can only be one active `close()` at a time.

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java [hunk 1]
```diff
@@ -1011,33 +1036,47 @@ public void completeIfEmpty() {
...
+            this.closeRequest = closeRequest;
         }
 
         public void setAsyncRequest(V asyncRequest) {
             this.asyncRequest = asyncRequest;
         }
 
-        public void setSyncRequest(V second) {
-            this.syncRequest = second;
+        public void setSyncRequestQueue(Queue<V> syncRequestQueue) {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: You could encapsulate this queue entirely in this class. For example, `addSyncRequest` could create the queue if it's not yet been created.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java [hunk 2]
```diff
@@ -84,10 +86,9 @@ public class ShareConsumeRequestManager implements RequestManager, MemberStateLi
     private Uuid memberId;
     private boolean fetchMoreRecords = false;
     private final Map<TopicIdPartition, Acknowledgements> fetchAcknowledgementsMap;
-    private final Map<Integer, Pair<AcknowledgeRequestState>> acknowledgeRequestStates;
+    private final Map<Integer, Tuple<AcknowledgeRequestState>> acknowledgeRequestStates;
     private final long retryBackoffMs;
     private final long retryBackoffMaxMs;
-    private boolean closing = false;
     private final CompletableFuture<Void> closeFuture;
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: It seems to me that the handling of `closeFuture` is flawed. An `AcknowledgeRequestState` is node-specific, but the future is associated with the entire request manager. It's only valid to complete the future once all close acks have been completed, not just the first one.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. A few comments from an initial review.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java (hunk 1 above)
Comment (ShivsundarR) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: Yeah makes sense, I have updated the addSyncRequest() method. Still the getSyncRequestQueue() method exists though as we need the queue contents when we process the acknowledgements.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java (hunk 2 above)
Comment (ShivsundarR) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: Yes, thanks, I had missed this. I am now put an extra variable to indicate if a requestState has completed processing, and then accordingly waited till all the nodes are finished and then completed the `closeFuture`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java (hunk 1 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/ShareConsumeRequestManager.java]: That's fine. It just seemed weird forcing the caller to make and set the queue.

**Review by AndrewJSchofield — APPROVED** [review]
lgtm. Thanks for the PR.

**Review by omkreddy — APPROVED** [review]

**Pull request merged**